### PR TITLE
[Fix #1409, #1407, #1424,#1442] Explicitly manage popwin 

### LINKED ISSUE #1409: helm-swoop scrolling is broken
I suspect this might be related to the recent helm pop-win change?

Before when I would scroll through the candidates in helm-swoop, the point in buffer would follow along, allowing one to see the context of the matches nicely. Now the other window gets scrolled instead.

### LINKED ISSUE #1424: Improve popwin
Since we can easily set the position of Helm window using built-in utility of Emacs like in this PR: https://github.com/syl20bnr/spacemacs/pull/1421, I think it's time we can drop `popwin`. You may ask, how can we control the position of temporary buffers like help or compilation buffer? Easy, do it like we do for Helm buffers:

```elisp
(add-to-list 'display-buffer-alist
               `(,(rx bos "*compilation")
                 (display-buffer-reuse-window
                  display-buffer-in-side-window)
                 (side . bottom)
                 (reusable-frames . visible)))

  (add-to-list 'display-buffer-alist
               `(,(rx bos "*help")
                 (display-buffer-reuse-window
                  display-buffer-in-side-window)
                 (side . bottom)
                 (reusable-frames . visible)))
```

You may ask, how can we delete these windows without popwin. The answer: we add a command for deleting these buffers, from anywhere:

```elisp
(defun lunaryorn-quit-bottom-side-windows ()
    "Quit side windows of the current frame."
    (interactive)
    (dolist (window (window-at-side-list))
      (when (not (string-match ".*helm.*" (buffer-name (window-buffer window))))
        (quit-window nil window))))

  (global-set-key (kbd "C-c [") 'lunaryorn-quit-bottom-side-windows)
```

(the command is taken from [here](https://github.com/lunaryorn/.emacs.d/commit/ab7669a34747d8b44ffa67787a94636472115652), with slight modification to not delete Helm windows)

If you want similar behavior to popwin - pressing `C-g` to quit the temporary window - then simply execute it before `keyboard-quit` runs:

```elisp
(advice-add 'keyboard-quit :before 'lunaryorn-quit-bottom-side-windows)
```

However, I think having a separate key binding is better since we may not always want such windows to be closed (i.e. `vc-diff` window).

Without `popwin`, it is much easier to integrate `window-purpose` and provide a much better window management experience.

### PR DESCRIPTION
**Pull request by tuhdo**
Set popwin:special-display-config to nil and let us manage it from
scratch. Doing so, we avoid popwin to mess up windows of some package
i.e. SLIME. Similar to how auto-completion layer is managed, if a layer
want to have popwin support, it must be explicitly specified in that
layer.

The default popwin:special-display-config contains obsolete config such
as anything (predecessor of Helm) and buffer that should be stick around
until we press C-g (like shell command output since we may need the
output for doing something else).

We don't need to use popwin:display-buffer since it can still mess up
window configuration, like this issue #1409. Simply add helm buffers to
display-buffer-alist, and use display-buffer-in-side-window; without any
argument, the default is bottom. We also inhibit Helm buffers to reuse
the existing windows with (inhibit-same-window . t) and a window
height (window-height . 0.4).

Also prevent popwin from piling up display-buffer-alist since everytime
it is reactivated, it adds its own buffers that already existed in
display-buffer-alist, by restoring display-buffer-alist after
popwin-mode is enabled again.

### CODE REVIEW

**Comment by tuhdo** [comment]
@syl20bnr you should really take a look at this.

**Comment by trishume** [comment]
:+1:

Pasting issue numbers in here so Github recognizes them: #1409, #1407, #1424,#1442

**Comment by mijoharas** [comment]
Thought I'd post it here as it is pretty relevant. I came across [this alternative to popwin](https://github.com/wasamasa/shackle) that could be useful, I haven't looked at the code much yet, and it might be useless in the light of this PR, but thought I'd point it out as we've been looking into this kind of stuff.

**Comment by tuhdo** [comment]
I already tested it before pushing this PR. It has problem with this issue: https://github.com/syl20bnr/spacemacs/issues/1396

To reproduce:

- Tell Shackle to manage Helm:

```elisp
(setq shackle-rules
      '(("\\`\\*helm.*?\\*\\'" :regexp t :align t :ratio 0.4)))
```

- Enable shackle: `M-x shackle-mode`.

- Run this code:

```elisp
(setq display-buffer-base-action '((display-buffer-use-some-window)))
```

- Now, `M-x` and press `TAB` to open help buffer. The help buffer is opened right inside Helm window and you cannot switch back to candidate buffer again with `TAB` or typing any character because this error

```elisp
Error running timer: (wrong-type-argument window-live-p nil) [5 times]
```

In other words, Helm window was killed, that's why it cannot be switched back.

**Comment by tuhdo** [comment]
@syl20bnr Here is a demo for @trishume case:

test

As you see, Helm always keeps the same height regardless how many windows there are.

**Comment by trishume** [comment]
:smile_cat:

**Comment by tuhdo** [comment]
@syl20bnr How about this PR then? It fixed many issues.

**Comment by syl20bnr** [comment]
@tuhdo I'm coming to it, be patient :-D

**Comment by syl20bnr** [comment]
Here it is :-)
Very nice PR :+1: :+1: :+1: 

Thank you ! 
Cherry-picked into `develop`, you can safely delete your branch.

**Comment by mijoharas** [comment]
:+1:
